File: core/safety/SafetyMetadataStore.cpp
#include "core/safety/SafetyMetadataStore.h"

#include <algorithm>
#include <chrono>
#include <cctype>
#include <fstream>
#include <iomanip>
#include <map>
#include <sstream>
#include <windows.h>

namespace gsm::core {
namespace {

std::string nowUtcIso8601()
{
    const auto now = std::chrono::system_clock::now();
    const std::time_t time = std::chrono::system_clock::to_time_t(now);
    std::tm utcTime{};
    gmtime_s(&utcTime, &time);

    std::ostringstream stream;
    stream << std::put_time(&utcTime, "%Y-%m-%dT%H:%M:%SZ");
    return stream.str();
}

std::string escapeValue(const std::string& value)
{
    std::string escaped;
    escaped.reserve(value.size());

    for (char character : value) {
        switch (character) {
        case '\\':
            escaped += "\\\\";
            break;
        case '\n':
            escaped += "\\n";
            break;
        case '\r':
            escaped += "\\r";
            break;
        case '|':
            escaped += "\\p";
            break;
        default:
            escaped += character;
            break;
        }
    }

    return escaped;
}

std::string unescapeValue(const std::string& value)
{
    std::string unescaped;
    unescaped.reserve(value.size());

    for (std::size_t index = 0; index < value.size(); ++index) {
        const char character = value[index];
        if (character != '\\' || index + 1 >= value.size()) {
            unescaped += character;
            continue;
        }

        const char next = value[++index];
        switch (next) {
        case 'n':
            unescaped += '\n';
            break;
        case 'r':
            unescaped += '\r';
            break;
        case 'p':
            unescaped += '|';
            break;
        case '\\':
            unescaped += '\\';
            break;
        default:
            unescaped += next;
            break;
        }
    }

    return unescaped;
}

std::vector<std::string> split(const std::string& value, char delimiter)
{
    std::vector<std::string> result;
    std::string current;
    std::istringstream stream(value);

    while (std::getline(stream, current, delimiter)) {
        result.push_back(current);
    }

    if (!value.empty() && value.back() == delimiter) {
        result.emplace_back();
    }

    return result;
}

std::string joinEscapedNotes(const std::vector<std::string>& notes)
{
    std::ostringstream stream;
    for (std::size_t index = 0; index < notes.size(); ++index) {
        if (index > 0) {
            stream << '|';
        }
        stream << escapeValue(notes[index]);
    }
    return stream.str();
}

std::vector<std::string> parseNotes(const std::string& notesText)
{
    std::vector<std::string> notes;
    if (notesText.empty()) {
        return notes;
    }

    for (const std::string& item : split(notesText, '|')) {
        notes.push_back(unescapeValue(item));
    }

    return notes;
}

std::optional<std::uintmax_t> parseUintMax(const std::string& value)
{
    try {
        return static_cast<std::uintmax_t>(std::stoull(value));
    } catch (...) {
        return std::nullopt;
    }
}

std::optional<std::uint64_t> parseUint64(const std::string& value)
{
    try {
        return static_cast<std::uint64_t>(std::stoull(value));
    } catch (...) {
        return std::nullopt;
    }
}

std::string fallbackNameFromPath(const std::string& path)
{
    const std::string normalized = gsm::system::normalizePath(path);
    const std::size_t separator = normalized.find_last_of("\\/");
    if (separator == std::string::npos) {
        return normalized;
    }
    return normalized.substr(separator + 1);
}

std::string toLower(std::string value)
{
    std::transform(value.begin(), value.end(), value.begin(), [](unsigned char character) {
        return static_cast<char>(std::tolower(character));
    });
    return value;
}

} // namespace

std::string toString(SafetyOperationState state)
{
    switch (state) {
    case SafetyOperationState::Planned:
        return "Planned";
    case SafetyOperationState::Running:
        return "Running";
    case SafetyOperationState::Completed:
        return "Completed";
    case SafetyOperationState::Failed:
        return "Failed";
    case SafetyOperationState::Restored:
        return "Restored";
    }

    return "Unknown";
}

std::optional<SafetyOperationState> parseSafetyOperationState(const std::string& value)
{
    const std::string lower = toLower(value);

    if (lower == "planned") {
        return SafetyOperationState::Planned;
    }
    if (lower == "running") {
        return SafetyOperationState::Running;
    }
    if (lower == "completed") {
        return SafetyOperationState::Completed;
    }
    if (lower == "failed") {
        return SafetyOperationState::Failed;
    }
    if (lower == "restored") {
        return SafetyOperationState::Restored;
    }

    return std::nullopt;
}

SafetyMetadataStore::SafetyMetadataStore(gsm::system::Path storageRoot)
    : storageRoot_(gsm::system::normalizePath(storageRoot))
{
}

SafetyMetadata SafetyMetadataStore::createPlannedMetadata(
    const GameAnalysis& analysis,
    const CompressionRecommendation& recommendation,
    const std::string& gameName,
    const std::string& source) const
{
    SafetyMetadata metadata;
    metadata.id = makeStableId(analysis.rootPath);
    metadata.gameName = gameName.empty() ? fallbackNameFromPath(analysis.rootPath) : gameName;
    metadata.rootPath = gsm::system::normalizePath(analysis.rootPath);
    metadata.source = source.empty() ? "manual" : source;
    metadata.sizeBeforeBytes = analysis.totalBytes;
    metadata.fileCountBefore = analysis.fileCount;
    metadata.algorithm = recommendation.algorithm;
    metadata.state = SafetyOperationState::Planned;
    metadata.createdAtUtc = nowUtcIso8601();
    metadata.updatedAtUtc = metadata.createdAtUtc;

    for (const std::string& reason : recommendation.reasons) {
        metadata.notes.push_back("recommendation:" + reason);
    }

    return metadata;
}

bool SafetyMetadataStore::save(const SafetyMetadata& metadata) const
{
    if (metadata.id.empty() || !gsm::system::ensureDirectoryExists(storageRoot_)) {
        return false;
    }

    std::ofstream file(metadataPathForId(metadata.id), std::ios::binary | std::ios::trunc);
    if (!file) {
        return false;
    }

    file << serializeSafetyMetadata(metadata);
    return file.good();
}

std::optional<SafetyMetadata> SafetyMetadataStore::loadById(const std::string& id) const
{
    std::ifstream file(metadataPathForId(id), std::ios::binary);
    if (!file) {
        return std::nullopt;
    }

    std::ostringstream buffer;
    buffer << file.rdbuf();
    return deserializeSafetyMetadata(buffer.str());
}

std::vector<SafetyMetadata> SafetyMetadataStore::loadAll() const
{
    std::vector<SafetyMetadata> records;
    const std::string searchPath = gsm::system::joinPath(storageRoot_, "*.gsmmeta");

    WIN32_FIND_DATAA findData;
    HANDLE findHandle = FindFirstFileA(searchPath.c_str(), &findData);
    if (findHandle == INVALID_HANDLE_VALUE) {
        return records;
    }

    do {
        const std::string fileName = findData.cFileName;
        const auto metadata = loadById(fileName.substr(0, fileName.find_last_of('.')));
        if (metadata.has_value()) {
            records.push_back(*metadata);
        }
    } while (FindNextFileA(findHandle, &findData) != 0);

    FindClose(findHandle);

    std::sort(records.begin(), records.end(), [](const SafetyMetadata& left, const SafetyMetadata& right) {
        return left.updatedAtUtc > right.updatedAtUtc;
    });

    return records;
}

gsm::system::Path SafetyMetadataStore::storageRoot() const
{
    return storageRoot_;
}

gsm::system::Path SafetyMetadataStore::metadataPathForId(const std::string& id) const
{
    return gsm::system::joinPath(storageRoot_, id + ".gsmmeta");
}

std::string SafetyMetadataStore::makeStableId(const std::string& rootPath)
{
    const std::string normalized = toLower(gsm::system::normalizePath(rootPath));
    std::uint64_t hash = 1469598103934665603ULL;

    for (unsigned char character : normalized) {
        hash ^= character;
        hash *= 1099511628211ULL;
    }

    std::ostringstream stream;
    stream << std::hex << hash;
    return stream.str();
}

std::string serializeSafetyMetadata(const SafetyMetadata& metadata)
{
    std::ostringstream stream;
    stream << "schemaVersion=" << metadata.schemaVersion << '\n';
    stream << "id=" << escapeValue(metadata.id) << '\n';
    stream << "gameName=" << escapeValue(metadata.gameName) << '\n';
    stream << "rootPath=" << escapeValue(metadata.rootPath) << '\n';
    stream << "source=" << escapeValue(metadata.source) << '\n';
    stream << "sizeBeforeBytes=" << metadata.sizeBeforeBytes << '\n';
    stream << "sizeAfterBytes=" << metadata.sizeAfterBytes << '\n';
    stream << "fileCountBefore=" << metadata.fileCountBefore << '\n';
    stream << "algorithm=" << (metadata.algorithm.has_value() ? toString(*metadata.algorithm) : "") << '\n';
    stream << "state=" << toString(metadata.state) << '\n';
    stream << "createdAtUtc=" << escapeValue(metadata.createdAtUtc) << '\n';
    stream << "updatedAtUtc=" << escapeValue(metadata.updatedAtUtc) << '\n';
    stream << "notes=" << joinEscapedNotes(metadata.notes) << '\n';
    return stream.str();
}

std::optional<SafetyMetadata> deserializeSafetyMetadata(const std::string& text)
{
    std::map<std::string, std::string> values;
    std::istringstream stream(text);
    std::string line;

    while (std::getline(stream, line)) {
        if (line.empty()) {
            continue;
        }

        const std::size_t separator = line.find('=');
        if (separator == std::string::npos) {
            return std::nullopt;
        }

        values[line.substr(0, separator)] = line.substr(separator + 1);
    }

    SafetyMetadata metadata;

    if (values.find("schemaVersion") == values.end() || values["schemaVersion"] != "1") {
        return std::nullopt;
    }

    metadata.schemaVersion = 1;
    metadata.id = unescapeValue(values["id"]);
    metadata.gameName = unescapeValue(values["gameName"]);
    metadata.rootPath = unescapeValue(values["rootPath"]);
    metadata.source = unescapeValue(values["source"]);

    const auto sizeBefore = parseUintMax(values["sizeBeforeBytes"]);
    const auto sizeAfter = parseUintMax(values["sizeAfterBytes"]);
    const auto fileCount = parseUint64(values["fileCountBefore"]);
    const auto state = parseSafetyOperationState(values["state"]);

    if (!sizeBefore.has_value() || !sizeAfter.has_value() || !fileCount.has_value() || !state.has_value()) {
        return std::nullopt;
    }

    metadata.sizeBeforeBytes = *sizeBefore;
    metadata.sizeAfterBytes = *sizeAfter;
    metadata.fileCountBefore = *fileCount;
    metadata.state = *state;
    metadata.createdAtUtc = unescapeValue(values["createdAtUtc"]);
    metadata.updatedAtUtc = unescapeValue(values["updatedAtUtc"]);
    metadata.notes = parseNotes(values["notes"]);

    if (!values["algorithm"].empty()) {
        metadata.algorithm = parseCompressionAlgorithm(values["algorithm"]);
        if (!metadata.algorithm.has_value()) {
            return std::nullopt;
        }
    }

    return metadata;
}

} // namespace gsm::core

File: game_storage_manager/core/safety/safety_metadata_store.py
from __future__ import annotations

import glob
from datetime import datetime, timezone
from pathlib import Path

from game_storage_manager.core.analyzer.game_analysis import GameAnalysis
from game_storage_manager.core.rules_engine.recommendation_engine import (
    CompressionRecommendation,
    algorithm_to_string,
    parse_compression_algorithm,
)
from game_storage_manager.core.safety.safety_metadata import (
    SafetyMetadata,
    SafetyOperationState,
    parse_safety_operation_state,
    state_to_string,
)
from game_storage_manager.system.filesystem.filesystem import (
    ensure_directory_exists,
    join_path,
    normalize_path,
)


def _now_utc_iso8601() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def _escape_value(value: str) -> str:
    result = []
    for ch in value:
        if ch == "\\":
            result.append("\\\\")
        elif ch == "\n":
            result.append("\\n")
        elif ch == "\r":
            result.append("\\r")
        elif ch == "|":
            result.append("\\p")
        else:
            result.append(ch)
    return "".join(result)


def _unescape_value(value: str) -> str:
    result = []
    i = 0
    while i < len(value):
        ch = value[i]
        if ch != "\\" or i + 1 >= len(value):
            result.append(ch)
            i += 1
            continue
        i += 1
        nxt = value[i]
        if nxt == "n":
            result.append("\n")
        elif nxt == "r":
            result.append("\r")
        elif nxt == "p":
            result.append("|")
        elif nxt == "\\":
            result.append("\\")
        else:
            result.append(nxt)
        i += 1
    return "".join(result)


def _join_escaped_notes(notes: list[str]) -> str:
    return "|".join(_escape_value(n) for n in notes)


def _parse_notes(notes_text: str) -> list[str]:
    if not notes_text:
        return []
    return [_unescape_value(item) for item in notes_text.split("|")]


def _fallback_name_from_path(path: str) -> str:
    normalized = normalize_path(path)
    normalized = normalized.replace("/", "\\")
    sep = normalized.rfind("\\")
    if sep == -1:
        return normalized
    return normalized[sep + 1:]


def _to_lower(value: str) -> str:
    return value.lower()


class SafetyMetadataStore:
    def __init__(self, storage_root: str):
        self._storage_root = normalize_path(storage_root)

    @property
    def storage_root(self) -> str:
        return self._storage_root

    def metadata_path_for_id(self, id: str) -> str:
        return join_path(self._storage_root, id + ".gsmmeta")

    def create_planned_metadata(
        self,
        analysis: GameAnalysis,
        recommendation: CompressionRecommendation,
        game_name: str,
        source: str,
    ) -> SafetyMetadata:
        metadata = SafetyMetadata()
        metadata.id = self.make_stable_id(analysis.root_path)
        metadata.game_name = game_name if game_name else _fallback_name_from_path(analysis.root_path)
        metadata.root_path = normalize_path(analysis.root_path)
        metadata.source = source if source else "manual"
        metadata.size_before_bytes = analysis.total_bytes
        metadata.file_count_before = analysis.file_count
        metadata.algorithm = recommendation.algorithm
        metadata.state = SafetyOperationState.Planned
        metadata.created_at_utc = _now_utc_iso8601()
        metadata.updated_at_utc = metadata.created_at_utc
        for reason in recommendation.reasons:
            metadata.notes.append("recommendation:" + reason)
        return metadata

    def save(self, metadata: SafetyMetadata) -> bool:
        if not metadata.id or not ensure_directory_exists(self._storage_root):
            return False
        try:
            path = self.metadata_path_for_id(metadata.id)
            with open(path, "w", encoding="utf-8", newline="") as f:
                f.write(serialize_safety_metadata(metadata))
            return True
        except OSError:
            return False

    def load_by_id(self, id: str) -> SafetyMetadata | None:
        try:
            path = self.metadata_path_for_id(id)
            with open(path, "r", encoding="utf-8") as f:
                return deserialize_safety_metadata(f.read())
        except OSError:
            return None

    def load_all(self) -> list[SafetyMetadata]:
        records: list[SafetyMetadata] = []
        search_pattern = join_path(self._storage_root, "*.gsmmeta")
        for file_path in glob.glob(search_pattern):
            file_name = Path(file_path).stem
            metadata = self.load_by_id(file_name)
            if metadata is not None:
                records.append(metadata)
        records.sort(key=lambda m: m.updated_at_utc, reverse=True)
        return records

    @staticmethod
    def make_stable_id(root_path: str) -> str:
        normalized = _to_lower(normalize_path(root_path))
        h = 1469598103934665603
        for ch in normalized:
            h ^= ord(ch)
            h = (h * 1099511628211) & 0xFFFFFFFFFFFFFFFF
        return format(h, "x")


def serialize_safety_metadata(metadata: SafetyMetadata) -> str:
    lines = [
        f"schemaVersion={metadata.schema_version}",
        f"id={_escape_value(metadata.id)}",
        f"gameName={_escape_value(metadata.game_name)}",
        f"rootPath={_escape_value(metadata.root_path)}",
        f"source={_escape_value(metadata.source)}",
        f"sizeBeforeBytes={metadata.size_before_bytes}",
        f"sizeAfterBytes={metadata.size_after_bytes}",
        f"fileCountBefore={metadata.file_count_before}",
        f"algorithm={algorithm_to_string(metadata.algorithm) if metadata.algorithm else ''}",
        f"state={state_to_string(metadata.state)}",
        f"createdAtUtc={_escape_value(metadata.created_at_utc)}",
        f"updatedAtUtc={_escape_value(metadata.updated_at_utc)}",
        f"notes={_join_escaped_notes(metadata.notes)}",
    ]
    return "\n".join(lines) + "\n"


def deserialize_safety_metadata(text: str) -> SafetyMetadata | None:
    values: dict[str, str] = {}
    for line in text.splitlines():
        if not line:
            continue
        sep = line.find("=")
        if sep == -1:
            return None
        values[line[:sep]] = line[sep + 1:]

    if values.get("schemaVersion") != "1":
        return None

    metadata = SafetyMetadata()
    metadata.schema_version = 1
    metadata.id = _unescape_value(values.get("id", ""))
    metadata.game_name = _unescape_value(values.get("gameName", ""))
    metadata.root_path = _unescape_value(values.get("rootPath", ""))
    metadata.source = _unescape_value(values.get("source", ""))

    try:
        metadata.size_before_bytes = int(values.get("sizeBeforeBytes", "0"))
        metadata.size_after_bytes = int(values.get("sizeAfterBytes", "0"))
        metadata.file_count_before = int(values.get("fileCountBefore", "0"))
    except ValueError:
        return None

    state = parse_safety_operation_state(values.get("state", ""))
    if state is None:
        return None
    metadata.state = state

    metadata.created_at_utc = _unescape_value(values.get("createdAtUtc", ""))
    metadata.updated_at_utc = _unescape_value(values.get("updatedAtUtc", ""))
    metadata.notes = _parse_notes(values.get("notes", ""))

    algo_str = values.get("algorithm", "")
    if algo_str:
        metadata.algorithm = parse_compression_algorithm(algo_str)
        if metadata.algorithm is None:
            return None

    return metadata
